### PR TITLE
added output dtype

### LINKED ISSUE #1240: Categorical cross entropy loss with integer vector
Hi,

Theano's `T.nnet.categorical_crossentropy` is a dual-behaviour function which accepts either a 2D tensor or an integer vector as the `true_dist` parameter.

I have a graph model with an output node that concatenates the output of N different softmax activations. I also have an integer vector representing the true labels.

By manually calling `T.nnet.categorical_crossentropy()` on my predictions and the true labels, I can obtain the loss. But this isn't working through Keras probably because of the `standardize_y()` function which forces the labels to be a 2D tensor. So we can't benefit of the second functionality of theano.

I implemented the following loss (which first flattens the true labels and then cast it to integer) to make things work:

```
def categorical_crossentropy_int(y_true, y_pred):
    return K.mean(K.categorical_crossentropy(y_pred, K.cast(y_true.flatten(), dtype='int32')), axis=-1)
```

What do you suggest as a fix for this? I saw that there's a utility function which converts integer label vector to one-hot encoded 2D tensor but I think with a huge class size (like 20k, 30k) this method is less efficient than the integer vector approach.

Thanks.

### PR DESCRIPTION
**Pull request by stevenxxiu**
Fixes #1240, #1391.

### CODE REVIEW

**Comment by fchollet** [comment]
- should be renamed to `output_dtype`
- dtypes being passed should be strings in `{float, int}`. These types are converted to float32 or float64 as appropriate. The type of float used in Keras is a global setting and the user is not expected to have to worry about it with every operation.
- user input should be validated and helpful error messages should be raised in case of invalid input.
- if this does solve the issues mentioned, then a simple test should be added to check that.

**Comment by stevenxxiu** [comment]
@fchollet Ok done, I'm raising an `Exception` if the user doesn't input a float or int.

**Comment by fchollet** [comment]
The API for Graph is not the most elegant, but it's acceptable. 

I believe the changes you made would break a number of features, such as model saving / loading. These changes are still untested.

**Comment by stevenxxiu** [comment]
@fchollet So do you want me to introduce some unit tests for model saving/loading, or are there already some which can be modified?

**Comment by fchollet** [comment]
I just want you to add unit tests for this feature (i.e. testing a small model that has a non-standard target format, like crossentropy with integers). These tests should include model loading / saving (take a look at `test_models.py` for examples).

BTW what is the justification for `output_ndim`? Shouldn't this always be inferred from the output ndim of the model itself?

**Comment by stevenxxiu** [comment]
> BTW what is the justification for output_ndim? Shouldn't this always be inferred from the output ndim of the model itself?

In the case of crossentropy it's one less, but I just wanted to leave some room for other cases where it's not (can't really think of any examples right now).

**Comment by stevenxxiu** [comment]
@fchollet Tests are added now.

**Comment by fchollet** [comment]
Obsolete PR, closing.
